### PR TITLE
[AutoDiff] simplify SILGen thunking and set correct thunk linkage

### PR DESCRIPTION
**Pull request by marcrasi**
Sets the SILGen'd AD thunk linkage to the same linkage as the custom derivative that they wrap, by doing two things:
* Use `customDerivativeFn->getLinkage()` for the linkage of the AD thunk.
* Use `ForDefinition` when getting the custom derivative functions. The `NotForDefinition` flag is for creating function declarations with external linkage, so if we use that then `customDerivativeFn->getLinkage()` is an external linkage, which is not right. (This code is probably going to be deleted very soon anyways, when we forbid custom derivatives in `@differentiable` attributes.)

Also it was initially hard for me to figure out what was going on, so I made some simplifications:
* Delete `getOrCreateAutoDiffDerivativeForwardingThunk` because we can just use the other thunking function, which has a superset of its functionality.
* Rename `getOrCreateAutoDiffDerivativeReabstractionThunk` -> `getOrCreateCustomDerivativeThunk` to clarify that it does all thunking necessary for custom derivatives, not just reabstraction.
* Make `getOrCreateCustomDerivativeThunk` determine whether self needs reordering, so that the caller does not have to worry about that.

### CODE REVIEW

**Review by dan-zheng — APPROVED** [review]
Sweet simplications!

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: lib/SILGen/SILGenThunk.cpp
Comment (dan-zheng) [inline lib/SILGen/SILGenThunk.cpp]: I believe as is, `getOrCreateAutoDiffDerivativeForwardingThunk` created a more efficient thunk, as it never reabstracts the returned linear map while `getOrCreateCustomDerivativeThunk` always does. Not a big deal!

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Review by rxwei — APPROVED** [review]

**Comment by marcrasi** [comment]
There's a failure in `validation-test/ParseableInterface/verify_all_overlays.py`. I think it's because my `ForDefinition` change is not quite right in certain cases involving `.swiftinterface` files. https://github.com/apple/swift/pull/28621 should make this problem go away, so I'll just wait until that gets merged and then retest this.

**Comment by dan-zheng** [comment]
Ping: https://github.com/apple/swift/pull/28621 has been merged, in case it unblocks this patch!

**Comment by marcrasi** [comment]
Oh, it seems that I also need https://bugs.swift.org/browse/TF-1001 for the problem to go away. So I'll wait until after I do that to retest this PR.

**Comment by marcrasi** [comment]
This has become unblocked. I've rebased it and will test it now.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Pull request merged**
